### PR TITLE
Loop in out nodes

### PR DESCRIPTION
**Pull request by vicdoval**
New nodes to create node loops. It seems very stable (much more than monads) although the inside logic is pretty similar, also pretty basic.
The loop is performed in the Loop Out node, this creates less interference with the actual update system.
Also they can do nested loops and use any viewer
image
image

- [x] Code changes complete.
- [x] Code documentation complete.
- [x] Documentation for users complete (or not required, if user never sees these changes).
- [x] Manual testing done. 
- [ ] Unit-tests implemented.
- [x] Ready for merge.

### CODE REVIEW

**Comment by portnov** [comment]
Will there be "for each" mode? Like, for each item in the input list, run the following nodes...

**Comment by vicdoval** [comment]
@Durman with the event system is possible to get from one node which nodes have been updated in the last update wave?

**Comment by vicdoval** [comment]
"For each" mode added :) 
image

**Comment by Durman** [comment]
Events do nothing. They are only means for communication with update system. I had added them for future. At this moment they are only used in `group` update system.

**Comment by Durman** [comment]
Debugging loops is the issue.

**Comment by Durman** [comment]
thing generator

**Comment by vicdoval** [comment]
Better Exceptions for debugging
image

File: nodes/logic/loop_out.py [hunk 1]
```diff
@@ -24,6 +24,12 @@
 from sverchok.core.update_system import make_tree_from_nodes, do_update
 from sverchok.data_structure import list_match_func
 
+def process_looped_nodes(node_list, tree_nodes):
+    for node_name in node_list:
+        try:
+            tree_nodes[node_name].process()
+        except Exception as e:
+            raise type(e)(str(e) + ' @ %s node' % node_name)
```
Comment (Durman) [inline nodes/logic/loop_out.py:32]: I think this syntax could be used.
https://docs.python.org/3/tutorial/errors.html#exception-chaining

Comment (vicdoval) [reply nodes/logic/loop_out.py]: That seems to change totally the exception and what we need is to add information to the existing exception https://stackoverflow.com/questions/6062576/adding-information-to-an-exception

**Comment by Durman** [comment]
Another idea. To add `while` mode and boolean socket to output loop node. Output node should looping until the socket will get False value.

**Comment by vicdoval** [comment]
**Break** input to break the loop
image
**Skip** input to skip that object in the output
image

**Comment by mnn** [comment]
I haven't tried it, but I am wondering, is it / would it be possible to name all those "data" input/output sockets (similarly like you can give custom names to sockets in group/monad)? I am not sure if it's that important feature, but it might make orientation in a node graph a bit easier - I wouldn't need to visually track `Data 0` to a node connecting to an input of a loop, if I would have appropriately named the socket in the loop node (e.g. seeing an output socket in the loop node named as  `HatVerts` means I don't have track from where `Data 0` is coming).

PS: Sorry for my English (those damn tenses), it's not my native language, but I hope it's understandable what I meant.

**Comment by vicdoval** [comment]
@mnn  that is a good point... i have just done a very simple implementation of that idea exposing `socket.label` in the properties 
of the Loop In node. The only downside is that the names dont spread (update) until the user creates a new link (or tries to)
image
@Durman you think it could be possible (and valuable) to being able to capture the changes in the socket.label from the owner node?

**Comment by Durman** [comment]
Probably it is possible but I don't think that it would be simplest solution. I would add code of changes of socket labels into `loop in` node. Then this code will be called each time new link will be connected to this node. Overhead of having such code in update method would not be noticeable I think.

**Comment by vicdoval** [comment]
I added a little operator to update the sockets label, looks less buggy image

File: docs/nodes/logic/loop_in.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Range mode example, Break used to control the maximum vertices.
+
+.. image:: https://user-images.githubusercontent.com/10011941/101332093-22234d00-3875-11eb-819a-68e86ef8c2c2.png
+
+For Each mode example, Skip input used to mask the results.
+
+.. image:: https://user-images.githubusercontent.com/10011941/101334215-e047d600-3877-11eb-89df-cfaaf73dd427.png
+
+You can change socket the socket labels in the N-Panel
```
Comment (mnn) [inline docs/nodes/logic/loop_in.rst:59]: First `socket` word looks redundant.

Comment (vicdoval) [reply docs/nodes/logic/loop_in.rst]: thanks!

File: nodes/logic/loop_in.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        self.inputs.new('SvStringsSocket', 'Iterations').prop_name = "iterations"
+        self.inputs.new('SvStringsSocket', 'Data 0')
+
+        self.outputs.new('SvLoopControlSocket', 'Loop Out')
+        self.outputs["Loop Out"].link_limit = 1
+        self.outputs.new('SvStringsSocket', 'Loop Number')
+        self.outputs.new('SvStringsSocket', 'Total Loops')
+
+    def draw_buttons(self,ctx, layout):
+        if not (self.outputs[0].is_linked and self.outputs[0].links[0].to_socket.node.bl_idname=='SvLoopOutNode'):
```
Comment (Durman) [inline nodes/logic/loop_in.py:124]: To call `socket.links` in draw method can be quite expensive.

File: nodes/logic/loop_in.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        if node.outputs[0].is_linked:
+            if node.outputs[0].links[0].to_socket.node.bl_idname == 'SvLoopOutNode':
+                loop_out_node = node.outputs[0].links[0].to_socket.node
+                for inp, self_inp, self_outp in zip(node.inputs[1:], loop_out_node.inputs[2:], loop_out_node.outputs):
+                    self_outp.label = inp.label
+                    self_inp.label = inp.label
+        return {'FINISHED'}
+
+
+class SvLoopInNode(bpy.types.Node, SverchCustomTreeNode):
```
Comment (Durman) [inline nodes/logic/loop_in.py:66]:
```suggestion
class SvLoopInNode(SverchCustomTreeNode, bpy.types.Node):
```

File: nodes/logic/loop_in.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    bl_idname = "node.create_loop_out"
+    bl_label = "Create Loop Out"
+
+    idtree: bpy.props.StringProperty(default='')
+    idname: bpy.props.StringProperty(default='')
+
+    def execute(self, context):
+        tree = bpy.data.node_groups[self.idtree]
+        node = bpy.data.node_groups[self.idtree].nodes[self.idname]
```
Comment (Durman) [inline nodes/logic/loop_in.py:36]:
```suggestion
    def execute(self, context):
        node = context.node
```

File: nodes/logic/loop_in.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        self.inputs.new('SvStringsSocket', 'Data 0')
+
+        self.outputs.new('SvLoopControlSocket', 'Loop Out')
+        self.outputs["Loop Out"].link_limit = 1
+        self.outputs.new('SvStringsSocket', 'Loop Number')
+        self.outputs.new('SvStringsSocket', 'Total Loops')
+
+    def draw_buttons(self,ctx, layout):
+        if not self.linked_to_loop_out:
+            self.wrapper_tracked_ui_draw_op(layout, "node.create_loop_out", icon='CON_FOLLOWPATH', text="Create Loop Out")
```
Comment (Durman) [inline nodes/logic/loop_in.py:129]:
```suggestion
            self.operator("node.create_loop_out", icon='CON_FOLLOWPATH', text="Create Loop Out")
```

**Comment by Durman** [comment]
It also would be useful to use loops for vectorization purposes. Because some times it is really hard to setup whole tree to proper vectorization.

**Comment by vicdoval** [comment]
The "For Each" mode can be used for this purpose: 
image

**Comment by portnov** [comment]
@vicdoval is there a way to pass different values of some parameter to internal node for each iteration? Like, extrude faces 3 times, first with height = 1.0, then 0.5 and then 0.25 ?

**Comment by vicdoval** [comment]
> @vicdoval is there a way to pass different values of some parameter to internal node for each iteration? Like, extrude faces 3 times, first with height = 1.0, then 0.5 and then 0.25 ?

like this?
image

**Comment by Durman** [comment]
@vicdoval 
 Ah, yes. This part is  a beat inconvenient only.

**Comment by vicdoval** [comment]
Its true the loop out node could be able to create its own sockets... But should this sockets appear in the "loop in node" too as outputs and inputs? that would requiere changing the actual socket logic...

**Comment by Durman** [comment]
Looks like I'm not understand `for each` mode. Should not the loop generate 5 boxes?

**Comment by vicdoval** [comment]
For each (Level 1) object
image

**Comment by Durman** [comment]
Ok. This looks is working. But this solution is about 6 times slower than similar solution without loops. Don't see why.

Next layout is 6 times faster.

**Pull request merged**